posts/inside-rust/2024-09-23-rtn-call-for-testing.md

```diff
@@ -0,0 +1,165 @@
+
+### An example
+
+Consider a trait `Foo` with a `method` that returns a type of `impl Future<Output = ()>`. We want to write a function that calls `method` and spawns the future on another thread:
+
+```rust
+fn spawn<T>(f: impl Future<Output = T> + Send + 'static) {}
+
+trait Foo {
+    fn method() -> impl Future<Output = ()>; // <-- RPITIT.
+}
+
+fn needs_sendable_future<T: Foo>()
+where
+    // How do we further restrict `T::method()` to be `Send + 'static`?
+{
+    spawn(T::method());
+    //~^ ERROR: `impl Future<Output = ()>` is not `Send`!
+}
+```
+
+Specifically, we may not want to restrict the *declaration* of `Foo`, since changing it in the declaration would restrict *all* implementations of `Foo`.
+
+```rust
+trait Foo {
+    fn method() -> impl Future<Output = ()> + Send + 'static;
+    //                                      ~~~~~~~~~~~~~~~~
+    //                                      Not what we want.
+}
+```
+
+So, on stable Rust, we have no way of expressing this restriction when using AFIT or RPITIT. In contrast, we can express this today if we were to use a GAT directly:
+
+```rust
+trait Foo {
+    type MethodFuture: Future<Output = ()>;
+    fn method() -> Self::MethodFuture;
+}
+
+fn needs_sendable_future<T: Foo>()
+where
+    // We can restrict this to only implementors of `Foo`
+    // whose `MethodFuture` is `Send + 'static`, so we can
+    // call `spawn` below:
+    T::MethodFuture: Send + 'static
+{
+    spawn(T::method());
+}
+```
+
+However, using GATs means that implementors of `Foo` have to write out the return type explicitly, `type MethodFuture = ...`, which doesn't ([yet](https://github.com/rust-lang/rust/pull/120700)) work if we have an anonymous, unnameable `Future` type!
+
+## The solution
+
+In [RFC 3654] we introduced return type notation (RTN). This will allow us to write `where` clause bounds that restrict the return types of functions and methods that use async fn in traits (AFIT) and return-position impl Trait in traits (RPITIT). Extending the example above, RTN lets us write:
+
+```rust
+fn needs_sendable_future<T: Foo>()
+where
+    T::method(..): Send + 'static // ✨
+{
+    spawn(T::method());
+    //~^ Works!
+}
+```
+
+## Restrictions
+
+Currently, RTN is only allowed for trait associated functions and methods with lifetime generics (not const or type generics) that use:
+
+* async fn in traits (AFIT)
+* return-position impl Trait in traits (RPITIT) where the impl Trait is the outermost return type, i.e. `-> impl Trait`, but not `-> Box<impl Trait>`
+
+These restrictions are described in further detail in [RFC 3654].
+
+## How do I help?
+
+We'd love for you to test out this new feature!
```

Comment:
```suggestion
We'd love for you to test out this feature on the latest Rust nightly compiler[^nightly].

[^nightly]: Make sure to run `rustup update nightly` or update your compiler with your system package manager, since the feature is very new.
```

Reply:
Perhaps?

Reply:
By "system package manager" do you mean `apt` or `yum` or whatever? Do those usually have nightly in them?

I like this change overall because it's more precise. Before it could be interpreted as if the feature were already on stable.

Reply:
> By "system package manager" do you mean apt or yum or whatever? Do those usually have nightly in them?

Well I'm just trying to cover "people who dont use rustup" here.

Reply:
I like that! It reminds me of "or wherever you get your podcasts."